docs: :memo: add a style guide for writing Python tests (#135)
seedcase-project/design: style/tests.qmd

Reviewer: Only other thing to add would be to include this link in the `_quarto.yml` file so it gets added to the website.

Author: Done!

---

> title: "Test style guide"

Reviewer: ```suggestion
title: "Writing Python tests"
```

Since it is in the style guide section, it doesn't need the style guide in the name.

> description: "Guidelines on how to write tests of our Python code"

Reviewer 2: ```suggestion
description: "Guidelines on how to write tests for our Python code"
```

---

{{< include /includes/_wip.qmd >}}

## What to write tests for

> Write to write tests for:

Reviewer: ```suggestion
Write tests for:
```

- Every function, class, and method.
- Every possible input and output.
- Every possible error within the function.

## How to write tests

### Test location

We use [pytest](https://docs.pytest.org/en/stable/) for testing our
Python code and keep the [tests outside the application
code](https://docs.pytest.org/en/stable/explanation/goodpractices.html#tests-outside-application-code)
in a `tests` directory, like so:

``` plaintext
sprout/
├── core/
│   ├── my_function.py
├── tests/
│   ├── test_my_function.py
```

### Test files

Make one test file per Python script and only test the functions within
that script. This makes it easier to do a cycle of writing and testing,
since you only need to run the one test file and not all the tests.

Always name the test file the same as the Python script, but with
`test_` as a prefix. E.g., `my_function.py` --\> `test_my_function.py`.

### Test names

Use the following general pattern for naming tests:

> - `test_<action>_<condition>`. E.g.,

Author: This is not taken from any framework or anything I have read on how to name tests, but rather just what it seems like we're (mostly) doing right now

Reviewer 3: I think this was inspired by Philip...

    `test_returns_relative_path_with_multi_digits` or
    `test_rejects_unsupported_format`.

> - If there's multiple functions in the same script, you can include

Reviewer 2: ```suggestion
- If there are multiple functions in the same script, you can include
```

    the function name in the test name:
    `test_<function>_<action>_<condition>`.

### Test structure

Use the following general structure for tests:

1. Arrange: Set up the test data and any other necessary conditions.
2. Act: Call the function or method being tested.
3. Assert: Check the output of the function or method.

> This is sometimes called "Given, When, Then".

Author: I noticed that @martonvago use this. I looked (quite quickly into it) and it seems to be very similar - or even the same? - as AAA? But please correct me if I'm wrong 😁

Reviewer 2: Yeah, same, afaik! 
(And yeah, I think it's okay that we're not religiously structuring the tests by always marking these stages with comments when the tests are short. But it can be nice to mark them "formally" when the tests get longer.)

Reviewer: I think I prefer the Given, when, then language since it is a bit "simpler" and more plain language compared to arrange, act, assert (all of which are not as commonly used words)
